Allow stack traces for warning logging

raboof opened the pull request:
I don't think LoggingAdapter was intended to be subclassed, right? We could
provide a default implementation to `notifyWarning(cause, message)` that just
calls `notifyWarning(message)` to be safer in terms of bincompat, but it
doesn't seem worth it.

raboof committed:
Allow stack traces for warning logging

raboof committed:
Add to JavaLogger as well

.. and add default impl for bincompat, as apparently
we've recommended subclassing LoggingAdapter :)

patriknw commented:
> I don't think LoggingAdapter was intended to be subclassed, right?

I don't think we can completely assume that. Probably it should have been final (DoNotInherit) but for historical reasons we might have to treat it as public api for extensions. That is probably also the reason why we have so many subclasses of it when we introduced new features.

raboof commented:
I've added a default implementation for all methods I added, which makes subclasses of `LoggingAdapter` not fail outright, but of course it's still strictly speaking not binary compatible.

ktoso commented on akka-actor/src/main/scala/akka/event/Logging.scala:
```diff
@@ -1770,6 +1824,7 @@ object NoMarkerLogging extends MarkerLoggingAdapter(null, "source", classOf[Stri
   final protected override def notifyError(message: String): Unit = ()
   final protected override def notifyError(cause: Throwable, message: String): Unit = ()
   final protected override def notifyWarning(message: String): Unit = ()
+  final protected override def notifyWarning(cause: Throwable, message: String): Unit = ()
```
this will break if a library extended LoggingAdapter and an user upgrades akka.

raboof replied:
Aw, forgot about the trait forwarders problem, yes :(

ktoso replied:
“This is why we can’t have nice things”

ktoso reviewed (changes requested):
No link to existing ticket? We're getting rather loose with hip-shot PRs recently a bit hm...

This exact change was rejected a few times due to binary compatibility issues. If we really really want this, it would have to be backed with researching that no widely used lib extends this and get a huge warning sign... In general not sure it is worth it; we have fixed it in Akka Typed's Logger though.

patriknw commented:
I agree that it's not worth the effort. One can use `Logging.stackTraceFor(e)` and an ordinary parameter if a stack trace is really needed in a warning.
